lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -1,23 +1,126 @@
+
+class SimpleIPRouteProviderCharm(ops.CharmBase):
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.RouterProvider = RouterProvides(charm=self)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.routing_table_updated, self._routing_table_updated)
+        self.framework.observe(self.on.get_routing_table_action, self._action_get_routing_table)
+
+    def _on_install(self, event: ops.InstallEvent):
+        pass
+
+    def _routing_table_updated(self, event):
+        routing_table = self.RouterProvider.get_routing_table()
+        all_networks = self.RouterProvider.get_flattened_routing_table()
+
+        # Process the networks however you like
+        implement_networks(all_networks)
+        
+        
+    def _action_get_routing_table(self, event: ops.ActionEvent):
+        all_networks = self.RouterProvider.get_flattened_routing_table()
+        event.set_results({"msg": json.dumps(all_networks)})
+    
+
+
+if __name__ == "__main__":  # pragma: nocover
+    ops.main(SimpleIPRouteProviderCharm)  # type: ignore
+```
+
+### Requirer charm
+This example requirer charm shows the two available actions as a host in the network:
+* get the latest list of all networks available from the provider
+* request a network to be assigned to the requirer charm
+
+The ip-router requirer allows a foolproof, typechecked, secure and safe way to 
+interact with the router that handles validation and format of the network 
+request, so you can focus on more important things. The library also provides a
+way to list out all of the available networks. This list is not cached, and comes
+directly from the provider.
+
+```python
+import logging, json
+import ops
+
+from charms.ip_router_interface.v0.ip_router_interface import *
+
+class SimpleIPRouteRequirerCharm(ops.CharmBase):
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.RouterRequirer = RouterRequires(charm=self)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.ip_router_relation_joined, self._on_relation_joined)
+
+        self.framework.observe(self.on.get_all_networks_action, self._action_get_all_networks)
+        self.framework.observe(self.on.request_network_action, self._action_request_network)
+
+    def _on_install(self, event: ops.InstallEvent):
+        pass
+
+    def _on_relation_joined(self, event: ops.RelationJoinedEvent):
+        self.unit.status = ops.ActiveStatus("Ready to Require")
+
+    def _action_get_all_networks(self, event: ops.ActionEvent):
+        # Get and process all of the available networks
+        all_networks = self.RouterRequirer.get_all_networks()
+        event.set_results({"msg": json.dumps(all_networks)})
+
+    def _action_request_network(self, event: ops.ActionEvent):
+        # Request a new network as required in the required format
```

Comment:
Shouldn't we have a mechanism to notify the requirer charm that networks are available? without expecting each requirer charm to implement these actions?

Reply:
I felt that it was a bit redundant since in the perspective of the requirer, the routing table is updated exactly when the ip-router_relation_changed fires, but I guess it wouldn't hurt to make it more explicit.